x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlConfigMigrator.java

```diff
@@ -342,6 +351,62 @@ public static Job updateJobForMigration(Job job) {
                 .collect(Collectors.toList());
     }
 
+    public static class JobsAndDatafeeds  {
+        List<Job> jobs;
+        List<DatafeedConfig> datafeedConfigs;
+
+        private JobsAndDatafeeds() {
+            jobs = new ArrayList<>();
+            datafeedConfigs = new ArrayList<>();
+        }
+
+        public int totalCount() {
+            return jobs.size() + datafeedConfigs.size();
+        }
+    }
+
+    /**
+     * Return at most {@link #MAX_BULK_WRITE_SIZE} configs favouring
+     * datafeed and job pairs so if a datafeed is choosen so is its job.
+     *
+     * @param datafeedsToMigrate Datafeed configs
+     * @param jobsToMigrate      Job configs
+     * @return Job and datafeed configs
+     */
+    public static JobsAndDatafeeds limitWrites(Collection<DatafeedConfig> datafeedsToMigrate, Map<String, Job> jobsToMigrate) {
+        JobsAndDatafeeds jobsAndDatafeeds = new JobsAndDatafeeds();
+
+        if (datafeedsToMigrate.size() + jobsToMigrate.size() <= MAX_BULK_WRITE_SIZE) {
+            jobsAndDatafeeds.jobs.addAll(jobsToMigrate.values());
+            jobsAndDatafeeds.datafeedConfigs.addAll(datafeedsToMigrate);
+            return jobsAndDatafeeds;
+        }
+
+        int count = 0;
+
+        // prioritise datafeed and job pairs
+        for (DatafeedConfig datafeedConfig : datafeedsToMigrate) {
+            if (count < MAX_BULK_WRITE_SIZE) {
+                Job datafeedsJob = jobsToMigrate.remove(datafeedConfig.getJobId());
+                jobsAndDatafeeds.jobs.add(datafeedsJob);
```

Comment:
This seems weird to me. It seems like you would risk adding `null` values to the list (unsure if you handle this down stream or not). Two concerns

* I am not why the `count` is being incremented on a possibly `null` value. 
* do we not care about datafeed configs that fit your `null` job scenario? As they are never added.

Reply:
The intention was there but the execution was poor. I made a mess of the logic have fixed that and added a test case covering the null job